storage/remote/queue_manager.go

```diff
@@ -1232,20 +1240,95 @@ func (s *shards) stop() {
 	}
 }
 
+func (s *shards) updateDroppedMetrics() {
+	// In this case we drop all samples in the buffer and the queue.
+	// Remove them from pending and mark them as failed.
+	droppedSamples := int(s.enqueuedSamples.Load())
+	droppedExemplars := int(s.enqueuedExemplars.Load())
+	droppedHistograms := int(s.enqueuedHistograms.Load())
+	s.qm.metrics.pendingSamples.Sub(float64(droppedSamples))
+	s.qm.metrics.pendingExemplars.Sub(float64(droppedExemplars))
+	s.qm.metrics.pendingHistograms.Sub(float64(droppedHistograms))
+	s.qm.metrics.failedSamplesTotal.Add(float64(droppedSamples))
+	s.qm.metrics.failedExemplarsTotal.Add(float64(droppedExemplars))
+	s.qm.metrics.failedHistogramsTotal.Add(float64(droppedHistograms))
+	s.samplesDroppedOnHardShutdown.Add(uint32(droppedSamples))
+	s.exemplarsDroppedOnHardShutdown.Add(uint32(droppedExemplars))
+	s.histogramsDroppedOnHardShutdown.Add(uint32(droppedHistograms))
+}
+
+func (s *shards) reshard(numShards int) bool {
+	// Exclusive lock to ensure that this does not run concurrently with enqueue.
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	newQueues := make([]*queue, numShards)
+	for i := 0; i < numShards; i++ {
+		newQueues[i] = newQueue(s.qm.cfg.MaxSamplesPerSend, s.qm.cfg.Capacity)
+	}
+
+	successful := s.redistributeSamples(newQueues)
+	// We are not able to redistribute, we can revert to the start/stop loop.
+	if !successful {
+		return false
+	}
+
+	// We have successfully moved all the samples, now can delete the old queues.
+	for _, queue := range s.queues {
+		close(queue.batchQueue)
+		queue.batchMtx.Unlock()
+	}
+
+	// Waiting till flushDeadline for all the runShards to terminate.
+	select {
+	case <-s.done:
+	case <-time.After(s.qm.flushDeadline):
+		// Cancelling the current context so as to unblock client calls.
+		s.hardShutdown()
+		<-s.done
+	}
+
+	s.queues = newQueues
+	var hardShutdownCtx context.Context
+	hardShutdownCtx, s.hardShutdown = context.WithCancel(context.Background())
+	s.running.Store(int32(numShards))
+	s.done = make(chan struct{})
+	for i := 0; i < numShards; i++ {
+		go s.runShard(hardShutdownCtx, i, newQueues[i])
+	}
```

Comment:
yeah potentially we want to start these immediately after creating the queues, otherwise we might fail to redistribute when downsharding